File: packages/sdk-java/src/main/java/dev/agentshield/AgentShield.java
package dev.agentshield;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.function.Function;

public class AgentShield {

    private final AgentShieldConfig config;
    private final PolicyEvaluator evaluator;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;

    public AgentShield(AgentShieldConfig config) {
        this.config = config != null ? config : new AgentShieldConfig();
        this.evaluator = new PolicyEvaluator();
        this.httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(5))
            .build();
        this.objectMapper = new ObjectMapper();
    }

    public EvaluationResult guard(ToolCallRequest request, GuardrailPolicy policy) {
        EvaluationResult result = evaluator.evaluate(request, policy);

        if (!result.isAllowed()) {
            if (config.getOnViolation() != null) {
                config.getOnViolation().accept(result, request);
            }
            String webhookUrl = policy.getWebhookUrl();
            if (webhookUrl == null || webhookUrl.isEmpty()) {
                webhookUrl = config.getWebhookUrl();
            }
            if (webhookUrl != null && !webhookUrl.isEmpty()) {
                dispatchWebhookAlert(result, request, webhookUrl);
            }
        }

        if (config.getTelemetryUrl() != null && !config.getTelemetryUrl().isEmpty()) {
            sendTelemetry(result, request);
        }

        return result;
    }

    public <T, R> Function<T, R> wrapTool(String toolName, Function<T, R> toolFn, GuardrailPolicy policy) {
        return (input) -> {
            ToolCallRequest request = new ToolCallRequest();
            request.setToolName(toolName);
            if (input instanceof Map) {
                request.setParams((Map<String, Object>) input);
            } else {
                try {
                    request.setParams(objectMapper.convertValue(input, Map.class));
                } catch (Exception e) {
                    request.setParams(Map.of("input", input));
                }
            }

            EvaluationResult evalResult = guard(request, policy);
            if (!evalResult.isAllowed()) {
                throw new AgentShieldException("[AgentShield Blocked] " + evalResult.getReason());
            }

            if (policy.getTimeoutMs() != null && policy.getTimeoutMs() > 0) {
                return executeWithTimeout(() -> toolFn.apply(input), policy.getTimeoutMs());
            }

            return toolFn.apply(input);
        };
    }

    private <R> R executeWithTimeout(java.util.concurrent.Callable<R> task, int timeoutMs) {
        try {
            return CompletableFuture.supplyAsync(() -> {
                try {
                    return task.call();
                } catch (Exception e) {
                    throw new RuntimeException(e);
                }
            }).get(timeoutMs, java.util.concurrent.TimeUnit.MILLISECONDS);
        } catch (java.util.concurrent.TimeoutException e) {
            throw new AgentShieldException("[AgentShield Timeout] Execution timed out after " + timeoutMs + "ms");
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private void dispatchWebhookAlert(EvaluationResult result, ToolCallRequest request, String webhookUrl) {
        try {
            Map<String, Object> payload = Map.of(
                "event", "AGENTSHIELD_VIOLATION_BLOCKED",
                "toolName", request.getToolName(),
                "reason", result.getReason(),
                "timestamp", result.getTimestamp(),
                "params", request.getParams()
            );
            String json = objectMapper.writeValueAsString(payload);

            HttpRequest httpRequest = HttpRequest.newBuilder()
                .uri(URI.create(webhookUrl))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();

            httpClient.sendAsync(httpRequest, HttpResponse.BodyHandlers.discarding());
        } catch (Exception ignored) {
            // Non-blocking background alert
        }
    }

    private void sendTelemetry(EvaluationResult result, ToolCallRequest request) {
        try {
            Map<String, Object> payload = Map.of(
                "agentId", request.getAgentId(),
                "toolName", request.getToolName(),
                "params", request.getParams(),
                "actionTaken", result.getActionTaken().name(),
                "reason", result.getReason(),
                "timestamp", result.getTimestamp()
            );
            String json = objectMapper.writeValueAsString(payload);

            HttpRequest httpRequest = HttpRequest.newBuilder()
                .uri(URI.create(config.getTelemetryUrl()))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();

            httpClient.sendAsync(httpRequest, HttpResponse.BodyHandlers.discarding());
        } catch (Exception ignored) {
            // Non-blocking telemetry sync
        }
    }

    public static class AgentShieldException extends RuntimeException {
        public AgentShieldException(String message) {
            super(message);
        }
    }
}

File: packages/sdk-dotnet/src/AgentShield/AgentShield.cs
using System.Net.Http.Json;
using System.Text.Json;

namespace AgentShield;

public class AgentShield
{
    private readonly AgentShieldConfig _config;
    private readonly PolicyEvaluator _evaluator;
    private readonly HttpClient _httpClient;

    public AgentShield(AgentShieldConfig? config = null)
    {
        _config = config ?? new AgentShieldConfig();
        _evaluator = new PolicyEvaluator();
        _httpClient = new HttpClient { Timeout = TimeSpan.FromSeconds(5) };
    }

    public EvaluationResult Guard(ToolCallRequest request, GuardrailPolicy policy)
    {
        EvaluationResult result = _evaluator.Evaluate(request, policy);

        if (!result.Allowed)
        {
            _config.OnViolation?.Invoke(result, request);

            string webhookUrl = policy.WebhookUrl ?? _config.WebhookUrl;
            if (!string.IsNullOrEmpty(webhookUrl))
            {
                _ = DispatchWebhookAlert(result, request, webhookUrl);
            }
        }

        if (!string.IsNullOrEmpty(_config.TelemetryUrl))
        {
            _ = SendTelemetry(result, request);
        }

        return result;
    }

    public Func<T, Task<R>> WrapTool<T, R>(string toolName, Func<T, Task<R>> toolFn, GuardrailPolicy policy)
    {
        return async (input) =>
        {
            var request = new ToolCallRequest
            {
                ToolName = toolName,
                Params = input is Dictionary<string, object> dict ? dict : ConvertToParams(input)
            };

            EvaluationResult evalResult = Guard(request, policy);
            if (!evalResult.Allowed)
            {
                throw new AgentShieldException($"[AgentShield Blocked] {evalResult.Reason}");
            }

            if (policy.TimeoutMs > 0)
            {
                using var cts = new CancellationTokenSource(TimeSpan.FromMilliseconds(policy.TimeoutMs));
                try
                {
                    return await toolFn(input).WaitAsync(cts.Token);
                }
                catch (OperationCanceledException)
                {
                    throw new AgentShieldException($"[AgentShield Timeout] Execution timed out after {policy.TimeoutMs}ms");
                }
            }

            return await toolFn(input);
        };
    }

    public Func<T, R> WrapToolSync<T, R>(string toolName, Func<T, R> toolFn, GuardrailPolicy policy)
    {
        return (input) =>
        {
            var request = new ToolCallRequest
            {
                ToolName = toolName,
                Params = input is Dictionary<string, object> dict ? dict : ConvertToParams(input)
            };

            EvaluationResult evalResult = Guard(request, policy);
            if (!evalResult.Allowed)
            {
                throw new AgentShieldException($"[AgentShield Blocked] {evalResult.Reason}");
            }

            return toolFn(input);
        };
    }

    private Dictionary<string, object> ConvertToParams<T>(T input)
    {
        try
        {
            return JsonSerializer.Deserialize<Dictionary<string, object>>(JsonSerializer.Serialize(input))
                ?? new Dictionary<string, object> { ["input"] = input };
        }
        catch
        {
            return new Dictionary<string, object> { ["input"] = input };
        }
    }

    private async Task DispatchWebhookAlert(EvaluationResult result, ToolCallRequest request, string webhookUrl)
    {
        try
        {
            var payload = new
            {
                event = "AGENTSHIELD_VIOLATION_BLOCKED",
                toolName = request.ToolName,
                reason = result.Reason,
                timestamp = result.Timestamp,
                @params = request.Params
            };
            await _httpClient.PostAsJsonAsync(webhookUrl, payload);
        }
        catch
        {
            // Non-blocking background alert
        }
    }

    private async Task SendTelemetry(EvaluationResult result, ToolCallRequest request)
    {
        try
        {
            var payload = new
            {
                agentId = request.AgentId,
                toolName = request.ToolName,
                @params = request.Params,
                actionTaken = result.ActionTaken.ToString(),
                reason = result.Reason,
                timestamp = result.Timestamp
            };
            await _httpClient.PostAsJsonAsync(_config.TelemetryUrl!, payload);
        }
        catch
        {
            // Non-blocking telemetry sync
        }
    }
}

public class AgentShieldException : Exception
{
    public AgentShieldException(string message) : base(message) { }
}